src/devices/Gpio/Drivers/VirtualGpioController.cs

```diff
@@ -0,0 +1,279 @@
+            {
+                throw new InvalidOperationException($"Pin {pinNumber} does not support mode {mode}.");
+            }
+
+            _pins[pinNumber].SetPinMode(mode);
+        }
+
+        /// <inheritdoc/>
+        public override void Write(int pinNumber, PinValue value)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not write to pin {pinNumber} because it is not open.");
+            }
+
+            _pins[pinNumber].Write(value);
+        }
+
+        /// <inheritdoc/>
+        public override void Toggle(int pinNumber)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not read from pin {pinNumber} because it is not open.");
+            }
+
+            _pins[pinNumber].Toggle();
+        }
+
+        /// <inheritdoc/>
+        public override void RegisterCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventTypes, PinChangeEventHandler callback)
+        {
+            if (_callbackEvents.Count == 0)
+            {
+                _pins[pinNumber].ValueChanged += PinValueChanged;
+            }
+
+            _callbackEvents.TryAdd(pinNumber, new PinEvent(eventTypes, callback));
+        }
+
+        private void PinValueChanged(object sender, PinValueChangedEventArgs pinValueChangedEventArgs)
+        {
+            // Get GpioPin associated with the real life pin
+            // Note that if we have multiple pins with the same pin number, we will only take the first one in the list
+            var pins = _pins.Where(m => m.Value.PinNumber == pinValueChangedEventArgs.PinNumber);
+            foreach (var pin in pins)
+            {
+                if (_callbackEvents.ContainsKey(pin.Key))
+                {
+                    var pinEvent = _callbackEvents[pin.Key];
+                    if (pinEvent == null)
+                    {
+                        return;
+                    }
+
+                    if (pinEvent.PinEventTypes == pinValueChangedEventArgs.ChangeType)
+                    {
+                        pinEvent.Callbacks?.Invoke(this, new PinValueChangedEventArgs(pinValueChangedEventArgs.ChangeType, pin.Key));
+                    }
+
+                    break;
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public override void UnregisterCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            _callbackEvents.TryRemove(pinNumber, out _);
+            if (_callbackEvents.Count == 0)
+            {
+                _pins[pinNumber].ValueChanged -= PinValueChanged;
+            }
+        }
+
+        /// <inheritdoc/>
+        public override WaitForEventResult WaitForEvent(int pinNumber, PinEventTypes eventTypes, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```

Comment:
This should be straight-forward, shouldn't it?

Reply:
no because I don't have access to the Driver neither. If we want to expose it, then yes, it can indeed be done

Reply:
Oh, got it. Would need the Driver property to be at least protected.

Reply:
I've been playing with this as well and it's not that straight forward. I may continue a bit but it's not easy and requires to create a virtual driver as well, things like this.